/*
    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        https://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
*/

syntax = "proto3";

import "SerializableDecomposedQueryPlan.proto";
import "SerializableOperator.proto";
import "google/protobuf/any.proto";

package NES;

/*
The serializable wrapper definition for query plan
 */
message SerializableReconfigurationMarker {
    map<string, SerializableReconfigurationMarkerEvent> reconfigurationMarkerEvents = 1;
}

message SerializableReconfigurationMarkerEvent {
    SerializableQueryState queryState = 1;
    SerializableReconfigurationMetadata reconfigurationMetadata = 2;
}

message SerializableNetworkSinkDescriptorUpdate {
    NodeLocation nodeLocation = 1;
    NesPartition nesPartition = 2;
    uint64 waitTime = 3;
    uint32 retryTimes = 4;
    uint64 uniqueNetworkSinkDescriptorId = 5;
    uint64 version = 6;
    uint64 numberOfOrigins = 7;
}

message SerializableReconfigurationMetadata {

    message SerializableDrainQueryMetadata {
        uint32 numberOfSources = 1;
    }

    message SerializableUpdateAndDrainMetadata {
        uint64 workerId = 1;
        uint64 sharedQueryId = 2;
        uint64 decomposedQueryId = 3;
        uint64 decomposedQueryPlanVersion = 4;
        uint32 numberOfSources = 5;
    }

    message SerializableUpdateQueryMetadata {
        uint64 workerId = 1;
        uint64 sharedQueryId = 2;
        uint64 decomposedQueryId = 3;
        uint64 decomposedQueryPlanVersion = 4;
        repeated SerializableNetworkSinkDescriptorUpdate sinkUpdates = 5;
    }

    google.protobuf.Any details = 1;
}